x-pack/solutions/observability/plugins/apm/public/components/shared/trace_waterfall/use_trace_waterfall.ts

```diff
@@ -95,6 +96,51 @@ export function getTraceParentChildrenMap(traceItems: TraceItem[]) {
   return traceMap;
 }
 
+export enum TraceDataState {
+  Full,
+  Partial,
+  Empty,
+}
+
+function matchParentId(this: TraceItem, parent: TraceItem) {
+  return parent.id === this.parentId;
+}
+
+function getFallbackRootItem(item: TraceItem, _index: number, traceItems: TraceItem[]): boolean {
+  // Avoid allocating a whole new function every iteration, making this operation less expensive
+  return !traceItems.find(matchParentId, item);
+}
+
+export function getRootItemOrFallback(
+  traceParentChildrenMap: Record<string, TraceItem[]>,
+  traceItems: TraceItem[]
+) {
+  const rootItem = traceParentChildrenMap.root?.[0];
+
+  if (rootItem) {
+    return {
+      traceState: TraceDataState.Full,
+      rootItem,
+    };
+  }
+
+  // O(n) in best case, O(n^2) in worst case. Hot loop here potentially.
+  // Avoid allocating new objects/functions here to ensure as fast iteration speed as
+  // possible for a potentially expensive op.
+  const fallbackItem = traceItems.find(getFallbackRootItem);
+
+  if (fallbackItem) {
```

Comment:
I'm not sure if I understood the logic here. Isn't it easier to check if the trace is empty and return `Partial` if it's not?

Reply:
If it is an empty list, it returns immediately as `undefined`, no? I don't think we _need_ to move the logic around here, as the empty case should not be expensive at all.

Reply:
Fair. To me it would be easier to understand if 

```ts
if (traces.length === 0) {
 return {
    traceState: TraceDataState.Empty,
  };
}

return {
   traceState: TraceDataState.Partial,
 }
 ```
 
 I'm not sure why we need to traverse the trace array, but maybe I'm missing some details